backend/cfg_selectgen.ml

```diff
@@ -897,59 +904,60 @@ module Make (Target : Cfg_selectgen_target_intf.S) = struct
         SU.set_traps_for_raise env;
         Ok rd
       | Terminator
-          (Prim { op = External ({ ty_args; ty_res; _ } as r); label_after }) ->
+          (Call
+            (External { func_symbol; ty_args; ty_res; alloc; returns; effects }))
+        -> (
         let* loc_arg, stack_ofs =
           emit_extcall_args env sub_cfg ty_args new_args dbg
         in
+        let function_really_returns = Option.is_some returns in
+        let ty_res_orig = ty_res in
+        let returns, ty_res =
+          match returns with
+          | Some _ -> returns, ty_res
+          | None ->
+            let keep_for_checking =
+              !SU.current_function_is_check_enabled
+              && String.equal func_symbol Cmm.caml_flambda2_invalid
+            in
+            if not keep_for_checking
+            then returns, ty
+            else Some (Cmm.new_label ()), Cmm.typ_int
+        in
         let rd = Reg.createv ty_res in
-        let term =
-          Cfg.Prim { op = External { r with stack_ofs }; label_after }
+        let term : Cfg.terminator =
+          Call
+            (External
+               { func_symbol;
+                 alloc;
+                 returns;
+                 effects;
+                 ty_res = ty_res_orig;
+                 ty_args;
+                 stack_ofs
+               })
         in
         let loc_res =
           SU.insert_op_debug' env sub_cfg term dbg loc_arg
             (Proc.loc_external_results (Reg.typv rd))
         in
-        Sub_cfg.add_never_block sub_cfg ~label:label_after;
         add_naming_op_for_bound_name sub_cfg loc_res;
-        SU.insert_move_results env sub_cfg loc_res rd stack_ofs;
         SU.set_traps_for_raise env;
-        Ok rd
-      | Terminator (Prim { op = Probe _; label_after } as term) ->
+        match returns with
+        | Some label_after ->
+          Sub_cfg.add_never_block sub_cfg ~label:label_after;
+          (* No need to insert result moves for [caml_flambda2_invalid] *)
+          if function_really_returns
+          then SU.insert_move_results env sub_cfg loc_res rd stack_ofs;
```

Comment:
The original code inserts the traps after the results move, the new code inserts traps before results move. I'm not sure if it matters.